### DISCUSSION TITLE
How to plot graph using the tool Add/Edit Tex on Xournal++

### DISCUSSION THREAD

**Discussion by chanemaster** [xournalpp/xournalpp#3144] [Q&A] [answered]
Hi, I now can type mathematical expression using the tool Add/Edit Tex on Xournal++.

However, I tried following Xournal++ guide below many times but not succeed to plot graph in Xournal++ by typing \graph{x^2} in the tool Add/Edit Tex. I don’t know much about LaTeX command. I just copied what the guide shown below to TeX file so it should not work properly.

Could you please give me some advice? Thank you in advance.

**The following is from Xournal++ Guide**

·        Include the PGFPlots package via \usepackage{pgfplots} for plotting function graphs. Using a macro you will be able to plot function graphs nicely and very quickly. For example, add the following lines to your template:

%for defining commands
\usepackage{xargs}
  
% for drawing and plotting
\usepackage{pgfplots}
\pgfplotsset{compat=newest} % Allows to place the legend below plot
\newcommandx{\graph}[3][1=,2=]{
    \begin{tikzpicture}
    \begin{axis}[xlabel=$x$,ylabel=$y$, axis lines=center,samples=100, #2]
    \addplot[#1]{ #3 };
    \end{axis}
    \end{tikzpicture}
}
Then a LaTeX-formula like \graph{x^2} will plot the normal parabola and \graph[domain=-2:2, blue][x=1cm,y=1cm]{x^2} will further restrict the domain, set the color and define units.

**Comment by rolandlo** [comment]
Can you post your full LaTeX preamble? When you attempt to create a LaTeX formula, you will find a `tex.tex` file somewhere in the temporary folder. Can you try to compile it manually with `pdflatex` and check for hints in the error message?

**Reply by chanemaster** [reply]
I named the tex file as “default_template.tex” and paste it under the folder C:\Users\MyName\AppData\Local\MiKTeX\TeXworks\0.6\templates\Basic LaTeX documents
(There are other 2 tex files name “article.tex” and “letter.tex” in this Basic LaTeX documents folder).

I tried to run “default_template” and read the error several times but I don’t understand how to fix it since I don’t know much about programming, The following is the full tex I used in "default_template.tex"

> \documentclass{article}
> \usepackage{tikz}
> 
> %for defining commands
> \usepackage{xargs}
> 
> % for drawing and plotting
> \usepackage{pgfplots}
> 
> \pgfplotsset{compat=newest} % Allows to place the legend below plot
> \newcommandx{\graph}[3][1=,2=]{
> \begin{tikzpicture}
> \begin{axis}[xlabel=$x$,ylabel=$y$, axis lines=center,samples=100, #2]
> \addplot[#1]{ #3 };
> \end{axis}
> \end{tikzpicture}
> }
> 
> \end{document}
>

**Reply by rolandlo** [reply]
> I named the tex file as “default_template.tex” and paste it under the folder C:\Users\MyName\AppData\Local\MiKTeX\TeXworks\0.6\templates\Basic LaTeX documents
> (There are other 2 tex files name “article.tex” and “letter.tex” in this Basic LaTeX documents folder).
> 
> I tried to run “default_template” and read the error several times but I don’t understand how to fix it since I don’t know much about programming, The following is the full tex I used in "default_template.tex"
> 
> > \documentclass{article}
> > \usepackage{tikz}
> > %for defining commands
> > \usepackage{xargs}
> > % for drawing and plotting
> > \usepackage{pgfplots}
> > \pgfplotsset{compat=newest} % Allows to place the legend below plot
> > \newcommandx{\graph}[3][1=,2=]{
> > \begin{tikzpicture}
> > \begin{axis}[xlabel=$x$,ylabel=$y$, axis lines=center,samples=100, #2]
> > \addplot[#1]{ #3 };
> > \end{axis}
> > \end{tikzpicture}
> > }
> > \end{document}

The idea was that you create a copy of the shipped out default template and make the modifications described in the guide. Please try this one: 
[template.zip](https://github.com/xournalpp/xournalpp/files/6602475/template.zip)

**Reply by chanemaster** [reply]
Thank you. I paste template.tex in the folder C:\Users\MyName\AppData\Local\MiKTeX\TeXworks\0.6\templates\Basic LaTeX documents but it still can not plot graph on Xournal++ using Add/Edit Tex tool.

So I tried running your template.tex and it showed this error:

> Runaway argument?
> {
> ! Paragraph ended before \XC@definec@lor was complete.
> <to be read again> 
>                    \par 
> l.17 
>      
> ?

**Reply by rolandlo** [reply]
> Thank you. I paste template.tex in the folder C:\Users\MyName\AppData\Local\MiKTeX\TeXworks\0.6\templates\Basic LaTeX documents but it still can not plot graph on Xournal++ using Add/Edit Tex tool.
> 
> So I tried running your template.tex and it showed this error:
> 
> > Runaway argument?
> > {
> > ! Paragraph ended before \XC@definec@lor was complete.
> > 
> > \par
> > l.17
> > ?

The template doesn't compile by itself. That's normal. Have you set the path of the global template path in Xournal++? menu `Edit > Preferences` LaTeX panel?

**Reply by chanemaster** [reply]
In my Xournal++, I clicked menu Edit > Preferences, I can't find menu to set the path of the global template. I also attached the screenshot for you.

![Capture](https://user-images.githubusercontent.com/85364302/120904215-4a99d180-c675-11eb-85d9-74a8b06256e5.PNG)

**Reply by rolandlo** [reply]
> In my Xournal++, I clicked menu Edit > Preferences, I can't find menu to set the path of the global template. I also attached the screenshot for you.
> 

The problem is clear now: You have version 1.0.20 (or older), but you need version 1.1.0+dev for this feature. Please download from the [nightly releases](https://github.com/xournalpp/xournalpp/releases/tag/nightly)

**Reply by chanemaster** [reply]
Thank you very much. I can plot graph already.
However, when the graph shown, it is usually appear like this screenshot below. I then have to press Enter only to plot graph but I cannot drag down this window or edit my typing anymore. Could you please tell me how to fix it? Thank you.

![Capture2](https://user-images.githubusercontent.com/85364302/120914261-22d35980-c6c7-11eb-8375-e651bfcc70fc.PNG)

**Comment by rolandlo** [comment, marked as answer]
The problem is, that the height of the graph is too big. Consider typing something like this:
```tex
\graph[][height=4cm]{x^2}
```
You could also set a default height in the LaTeX preamble if you encounter this problem repeatedly.

**Reply by chanemaster** [reply]
Thank you. Can I also define range of y to make the height smaller?

In the screenshot below, I used \graph[domain=-3:3,blue][x=1cm,y=1cm,height=4cm]{x^2}
and the range of y appear to be 0 <y < 9

I tried using \graph[domain=-3:3,range=-3:3,blue][x=1cm,y=1cm,height=4cm]{x^2}
but it does not work. So I guess the command range=-3:3 is wrong. May I ask you what is the right command to use for range of y, please?


![Capture3](https://user-images.githubusercontent.com/85364302/120916335-3edcf800-c6d3-11eb-96eb-0ef2d49162f4.PNG)

**Reply by rolandlo** [reply]
You can use the arguments `ymin=-3, ymax=3` instead of `range=-3:3`. See the [pgfplots manual](https://www.ctan.org/pkg/pgfplots) for more options.

**Reply by chanemaster** [reply]
Thank you for your assistance. This is nice. I can plot the graph and define range of y already.
Now I would like to save and export as pdf but they both are not working and show error message below.
Could you please help me how to save and export to pdf?

![save problem](https://user-images.githubusercontent.com/85364302/120931146-905ca580-c71a-11eb-8596-d3c568da96bf.PNG)
![Export as pdf problem](https://user-images.githubusercontent.com/85364302/120931150-95215980-c71a-11eb-92dc-7af9199a93a7.PNG)

**Reply by rolandlo** [reply]
> Now I would like to save and export as pdf but they both are not working and show error message below.
> Could you please help me how to save and export to pdf?

Hmm, I have never seen these error messages. Where (in which folder and under which name) are you trying to save the file? Looks like it is a system folder. Try some different folder.
I thought it might be a regression of the recently merged commit a5ea370ba14167c77114cb72af00e9261d7382f5, but I can't reproduce this file system error with the latest nightly release.

**Reply by chanemaster** [reply]
The error message shows everytime after I click File > Save, File > Export as PDF (also File > Save as, File > Export as...). So I am not be able to save any file at all. I also captured the full screenshot for you below.

![Export as pdf problem2](https://user-images.githubusercontent.com/85364302/120933770-c6ebed80-c725-11eb-8cfc-ca9ebd0f315d.PNG)

**Reply by rolandlo** [reply]
Ok, I see. Can you try to find the `settings.xml` file (in `C:\Users\YourUserName\AppData\Local\Xournal++\` or similar) and delete or rename it (which forces generation of a fresh settings file)? Maybe the `lastSavePath` stored in it is corrupted.

**Reply by chanemaster** [reply]
Hi, I changed the name of settings.xml and the file still cannot be saved. Then I went back to see the folder you mentioned 
 there is new settings.xml created again.

I was wondering about the error message:

> The folder contents could not be displayed
> 
> Error when getting information for file
> 
> “C:\Program Files\Xournal++\file:\C:\Program%20Files\Xournal++”
> 
> : Input/output error

Why does the path has C:\ and C:\ again which should not be possible, right?

**Reply by rolandlo** [reply]
Right, that should not be possible. I am wondering where it comes from, if not from the `settings.xml` file. Could you please file a [bug report](https://github.com/xournalpp/xournalpp/issues/new/choose)? We are getting off topic (graph plotting).

**Reply by chanemaster** [reply]
Okay, I submitted the bug already. Thank you for your kind assistance.
